test-data-generator/src/main/scala/za/co/absa/spline/testdatagen/NumericValue.scala

```diff
@@ -20,23 +20,34 @@ sealed trait NumericValue {
   def isExpanded: Boolean = this.isInstanceOf[Constant]
 
   def valueOf(): Int = if (isExpanded) this.asInstanceOf[Constant].value else {
-    throw new Exception("Not expanded")
+    throw new NotImplementedError("Not expanded")
   }
+
+  override def toString: String = valueOf().toString
 }
 case class Constant(value: Int) extends NumericValue
 
 case class Variable(start: Int, end: Int, step: Int) extends NumericValue {
   def expand(): Seq[Constant] = (start to end by step).map(i => Constant(i))
+
+  override def toString: String = s"${start}-${end}|${step}"
 }
 
 object NumericValue {
+  private val numberPattern = "([0-9]+)"
+
   def apply(param: String): NumericValue = {
-    val ConstantPattern = "([0-9]+)".r
-    val VariablePattern = "([0-9]+)-([0-9]+)/([0-9]+)".r
+    val ConstantPattern = numberPattern.r
+    val VariablePattern = (numberPattern + "-" + numberPattern + "/" + numberPattern).r
```

Comment:
I would make these these two vars constants, and inlined the `numberPattern`.
A little extra characters don't hurt too much in this case, but make both regexps literals that are much more readable than string concatenated ones. If you want to save on characters you can use the `\d` metamatcher.
```suggestion
    val ConstantPattern = """(\d+)""".r
    val VariablePattern = """(\d+)-(\d+)/(\d+)""".r
```